File: python_archive/src/models/session.py
"""Game session data model."""
from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, Any, Optional


@dataclass
class GameSession:
    """Represents a single game session for a specific mode."""
    mode: str
    start_time: datetime = field(default_factory=datetime.now)
    end_time: Optional[datetime] = None

    # Task completion tracking
    tasks_completed: int = 0
    current_streak: int = 0
    best_streak: int = 0

    # Accuracy tracking
    total_keystrokes: int = 0
    correct_keystrokes: int = 0
    errors: int = 0

    # XP earned this session
    xp_earned: int = 0

    # Mode-specific data (flexible storage)
    mode_data: Dict[str, Any] = field(default_factory=dict)

    # Exit status
    exited: bool = False

    def start(self):
        """Start the session."""
        self.start_time = datetime.now()
        self.exited = False

    def end(self):
        """End the session."""
        self.end_time = datetime.now()
        self.exited = True

    def duration_seconds(self) -> float:
        """
        Calculate session duration in seconds.

        Returns:
            Duration in seconds (0 if not ended)
        """
        if self.end_time is None:
            # Session still ongoing
            return (datetime.now() - self.start_time).total_seconds()
        return (self.end_time - self.start_time).total_seconds()

    def add_task_completion(self, xp: int):
        """
        Record a completed task.

        Args:
            xp: XP earned for this task
        """
        self.tasks_completed += 1
        self.xp_earned += xp
        self.current_streak += 1

        if self.current_streak > self.best_streak:
            self.best_streak = self.current_streak

    def break_streak(self):
        """Break the current streak (e.g., on error or exit)."""
        self.current_streak = 0

    def record_keystroke(self, correct: bool):
        """
        Record a keystroke.

        Args:
            correct: Whether the keystroke was correct
        """
        self.total_keystrokes += 1
        if correct:
            self.correct_keystrokes += 1
        else:
            self.errors += 1

    def calculate_accuracy(self) -> float:
        """
        Calculate accuracy percentage.

        Returns:
            Accuracy as a percentage (0-100), or 100 if no keystrokes
        """
        if self.total_keystrokes == 0:
            return 100.0
        return (self.correct_keystrokes / self.total_keystrokes) * 100

    def calculate_error_rate(self) -> float:
        """
        Calculate error rate percentage.

        Returns:
            Error rate as a percentage (0-100)
        """
        if self.total_keystrokes == 0:
            return 0.0
        return (self.errors / self.total_keystrokes) * 100

    def get_mode_data(self, key: str, default: Any = None) -> Any:
        """
        Get mode-specific data.

        Args:
            key: Key to retrieve
            default: Default value if key not found

        Returns:
            Value associated with key, or default
        """
        return self.mode_data.get(key, default)

    def set_mode_data(self, key: str, value: Any):
        """
        Set mode-specific data.

        Args:
            key: Key to set
            value: Value to store
        """
        self.mode_data[key] = value

    def __str__(self) -> str:
        """String representation."""
        duration = self.duration_seconds()
        return (
            f"GameSession({self.mode}, "
            f"{self.tasks_completed} tasks, "
            f"{self.calculate_accuracy():.1f}% accuracy, "
            f"{duration:.1f}s)"
        )

File: lua/typing_kata/core/session.lua
-- Game session tracking
local M = {}

-- Create a new game session
function M.new(mode_name)
  return {
    mode_name = mode_name,
    start_time = os.time(),
    end_time = nil,

    -- Task tracking
    tasks_completed = 0,
    current_streak = 0,
    best_streak = 0,

    -- Keystroke tracking
    total_keystrokes = 0,
    correct_keystrokes = 0,
    error_count = 0,

    -- XP tracking
    xp_earned = 0,

    -- Mode-specific data (flexible storage)
    mode_data = {},
  }
end

-- Record a keystroke
function M.record_keystroke(session, correct)
  session.total_keystrokes = session.total_keystrokes + 1

  if correct then
    session.correct_keystrokes = session.correct_keystrokes + 1
  else
    session.error_count = session.error_count + 1
    session.current_streak = 0  -- Break streak on error
  end
end

-- Calculate current accuracy
function M.calculate_accuracy(session)
  if session.total_keystrokes == 0 then
    return 100
  end
  return (session.correct_keystrokes / session.total_keystrokes) * 100
end

-- Add task completion
function M.add_task_completion(session, xp)
  session.tasks_completed = session.tasks_completed + 1
  session.xp_earned = session.xp_earned + xp
end

-- Increment streak
function M.increment_streak(session)
  session.current_streak = session.current_streak + 1
  if session.current_streak > session.best_streak then
    session.best_streak = session.current_streak
  end
end

-- Break streak (reset to 0)
function M.break_streak(session)
  session.current_streak = 0
end

-- End session
function M.end_session(session)
  session.end_time = os.time()
  return session
end

-- Get session duration in seconds
function M.get_duration(session)
  local end_time = session.end_time or os.time()
  return end_time - session.start_time
end

-- Get session summary as string
function M.get_summary(session)
  local duration = M.get_duration(session)
  local accuracy = M.calculate_accuracy(session)

  local lines = {
    '========================================',
    '          SESSION COMPLETE',
    '========================================',
    '',
    'Mode: ' .. session.mode_name,
    'Tasks Completed: ' .. session.tasks_completed,
    'XP Earned: ' .. session.xp_earned,
    '',
    'Accuracy: ' .. string.format('%.1f%%', accuracy),
    'Best Streak: ' .. session.best_streak,
    'Duration: ' .. string.format('%dm %ds', math.floor(duration / 60), duration % 60),
    '',
    'Total Keystrokes: ' .. session.total_keystrokes,
    'Correct: ' .. session.correct_keystrokes,
    'Errors: ' .. session.error_count,
    '========================================',
  }

  return table.concat(lines, '\n')
end

return M
